cmd/format-fs.go

```diff
@@ -0,0 +1,231 @@
+/*
+ * Minio Cloud Storage, (C) 2017 Minio, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package cmd
+
+import (
+	"encoding/json"
+	"fmt"
+	"io"
+	"os"
+	"time"
+
+	errors2 "github.com/minio/minio/pkg/errors"
+	"github.com/minio/minio/pkg/lock"
+)
+
+// FS format version strings.
+const (
+	formatBackendFS   = "fs"
+	formatFSVersionV1 = "1"
+)
+
+// formatFSV1 - structure holds format version '1'.
+type formatFSV1 struct {
+	formatMetaV1
+	FS struct {
+		Version string `json:"version"`
+	} `json:"fs"`
+}
+
+// Used to detect the version of "fs" format.
+type formatFSVersionDetect struct {
+	FS struct {
+		Version string `json:"version"`
+	} `json:"fs"`
+}
+
+// Returns the latest "fs" format.
+func newFormatFSV1() (format *formatFSV1) {
+	f := &formatFSV1{}
+	f.Version = formatMetaVersionV1
+	f.Format = formatBackendFS
+	f.FS.Version = formatFSVersionV1
+	return f
+}
+
+// Load the format.json
+func formatFSLoad(r io.ReadSeeker, data interface{}) error {
```

Comment:
Maybe we can see how to do this part better? The function is no different from a json decoder that starts with a io.ReadSeeker, and converts it into a io.Reader and loads up an arbitrary structure.

Reply:
it does 2 things:
1)  r.Seek(0, io.SeekStart)
2) decode

If we don't have it as a function we will have to repeat the above steps whenever we want to read the file (currently at 2 places). Also it is easy to forget to seek to 0. Hence it's better to have it as a function.

Reply:
Yes, but what I mean is, we could make the function more useful? E.g:

```go
func decodeJSONFromSeeker(r io.ReadSeeker, offset int64, data interface{}) error {
    	if _, err := r.Seek(offset, io.SeekStart); err != nil {
		return err
	}
	return json.NewDecoder(r).Decode(data)
}
```

And we could put it `cmd/utils.go`.

Reply:
done